File: lib/beetle/redis_server_info.rb
module Beetle
  # class used by the RedisConfigurationServer to hold information about the current state of the configured redis servers
  class RedisServerInfo
    include Logging

    def initialize(config, options) #:nodoc:
      @config = config
      @options = options
      reset
    end

    # all configured redis servers
    def instances
      @instances ||= @config.redis_servers.split(/ *, */).map{|s| Redis.from_server_string(s, @options)}
    end

    # fetches the server from the insatnces whith the given <tt>server</tt> string
    def find(server)
      instances.find{|r| r.server == server}
    end

    # refresh connectivity/role information
    def refresh
      logger.debug "Updating redis server info"
      reset
      instances.each {|r| @server_info[r.role] << r}
    end

    # subset of instances which are masters
    def masters
      @server_info["master"]
    end

    # subset of instances which are slaves
    def slaves
      @server_info["slave"]
    end

    # subset of instances which are not reachable
    def unknowns
      @server_info["unknown"]
    end

    # subset of instances which are set up as slaves of the given <tt>master</tt>
    def slaves_of(master)
      slaves.select{|r| r.slave_of?(master.host, master.port)}
    end

    # determine a master if we have one master and all other servers are slaves
    def auto_detect_master
      return nil unless master_and_slaves_reachable?
      masters.first
    end

    private

    def master_and_slaves_reachable?
      masters.size == 1 && slaves.size == instances.size - 1
    end

    def reset
      @server_info = Hash.new {|h,k| h[k]= []}
    end
  end
end

File: go/src/github.com/xing/beetle/redis_server_info.go
package main

import (
	"regexp"

	//	"github.com/davecgh/go-spew/spew"
	"gopkg.in/redis.v5"
)

type RedisShims []*RedisShim

func (shims RedisShims) Include(r *RedisShim) bool {
	for _, x := range shims {
		if x.server == r.server {
			return true
		}
	}
	return false
}

func (shims RedisShims) Servers() []string {
	servers := make([]string, 0)
	for _, x := range shims {
		servers = append(servers, x.server)
	}
	return servers
}

type RedisServerInfo struct {
	instances  RedisShims
	serverInfo map[string]RedisShims
}

func NewRedisServerInfo(servers string) *RedisServerInfo {
	si := &RedisServerInfo{}
	si.instances = make(RedisShims, 0)
	if servers != "" {
		serverList := regexp.MustCompile(" *, *").Split(servers, -1)
		for _, s := range serverList {
			logInfo("adding redis server: %s", s)
			si.instances = append(si.instances, NewRedisShim(s))
		}
	}
	si.Reset()
	return si
}

func (si *RedisServerInfo) NumServers() int {
	return len(si.instances)
}

func (si *RedisServerInfo) Reset() {
	m := make(map[string]RedisShims)
	m[MASTER] = make(RedisShims, 0)
	m[SLAVE] = make(RedisShims, 0)
	m[UNKNOWN] = make(RedisShims, 0)
	si.serverInfo = m
}

func (si *RedisServerInfo) Refresh() {
	logInfo("refreshing server info")
	si.Reset()
	for _, ri := range si.instances {
		role := ri.Role()
		logInfo("determined %s to be a '%s'", ri.server, role)
		si.serverInfo[role] = append(si.serverInfo[role], ri)
	}
	// spew.Dump(si)
}

func (si *RedisServerInfo) Find(server string) *redis.Client {
	for _, r := range si.instances {
		if r.server == server {
			return r.redis
		}
	}
	return nil
}

func (si *RedisServerInfo) Masters() RedisShims {
	return si.serverInfo[MASTER]
}

func (si *RedisServerInfo) Slaves() RedisShims {
	return si.serverInfo[SLAVE]
}

func (si *RedisServerInfo) Unknowns() RedisShims {
	return si.serverInfo[UNKNOWN]
}

func (si *RedisServerInfo) SlavesOf(master *RedisShim) RedisShims {
	slaves := make(RedisShims, 0)
	for _, s := range si.Slaves() {
		if s.IsSlaveOf(master.host, master.port) {
			slaves = append(slaves, s)
		}
	}
	return slaves
}

func (si *RedisServerInfo) AutoDetectMaster() *RedisShim {
	if !si.MasterAndSlavesReachable() {
		return nil
	}
	return si.Masters()[0]
}

func (si *RedisServerInfo) MasterAndSlavesReachable() bool {
	return len(si.Masters()) == 1 && len(si.Slaves()) == len(si.instances)-1
}
